include/rmm/mr/device/aligned_resource_adaptor.hpp

```diff
@@ -0,0 +1,221 @@
+   */
+  Upstream* get_upstream() const noexcept { return upstream_; }
+
+  /**
+   * @copydoc rmm::mr::device_memory_resource::supports_streams()
+   */
+  [[nodiscard]] bool supports_streams() const noexcept override
+  {
+    return upstream_->supports_streams();
+  }
+
+  /**
+   * @brief Query whether the resource supports the get_mem_info API.
+   *
+   * @return bool true if the upstream resource supports get_mem_info, false otherwise.
+   */
+  [[nodiscard]] bool supports_get_mem_info() const noexcept override
+  {
+    return upstream_->supports_get_mem_info();
+  }
+
+ private:
+  static constexpr std::size_t default_allocation_alignment = 256;
+  static constexpr std::size_t default_alignment_threshold  = 0;
+  using lock_guard                                          = std::lock_guard<std::mutex>;
+
+  /**
+   * @brief Allocates memory of size at least `bytes` using the upstream resource with the specified
+   * alignment.
+   *
+   * @throws `rmm::bad_alloc` if the requested allocation could not be fulfilled
+   * by the upstream resource.
+   *
+   * @param bytes The size, in bytes, of the allocation
+   * @param stream Stream on which to perform the allocation
+   * @return void* Pointer to the newly allocated memory
+   */
+  void* do_allocate(std::size_t bytes, cuda_stream_view stream) override
+  {
+    if (allocation_alignment_ == default_allocation_alignment || bytes < alignment_threshold_) {
+      return upstream_->allocate(bytes, stream);
+    } else {
+      auto const size            = upstream_allocation_size(bytes);
+      void* pointer              = upstream_->allocate(size, stream);
+      auto const address         = reinterpret_cast<std::size_t>(pointer);
+      auto const aligned_address = rmm::detail::align_up(address, allocation_alignment_);
+      void* aligned_pointer      = reinterpret_cast<void*>(aligned_address);
+      if (pointer != aligned_pointer) {
+        lock_guard lock(mtx_);
+        pointers_.emplace(aligned_pointer, pointer);
+      }
+      return aligned_pointer;
+    }
+  }
+
+  /**
+   * @brief Free allocation of size `bytes` pointed to to by `p` and log the deallocation.
+   *
+   * @throws Nothing.
+   *
+   * @param p Pointer to be deallocated
+   * @param bytes Size of the allocation
+   * @param stream Stream on which to perform the deallocation
+   */
+  void do_deallocate(void* p, std::size_t bytes, cuda_stream_view stream) override
+  {
+    if (allocation_alignment_ == default_allocation_alignment || bytes < alignment_threshold_) {
+      upstream_->deallocate(p, bytes, stream);
+    } else {
+      {
+        lock_guard lock(mtx_);
+        auto const i = pointers_.find(p);
+        if (i != pointers_.end()) {
+          p = i->second;
+          pointers_.erase(i);
+        }
+      }
+      upstream_->deallocate(p, upstream_allocation_size(bytes), stream);
+    }
+  }
+
+  /**
+   * @brief Compare the upstream resource to another.
```

Comment:
This doesn't just compare the upstream resource. It compares this resource to another.
```suggestion
   * @brief Compare this resource to another.
```

Reply:
Done.